src/browser/fake_network.js

```diff
@@ -796,67 +950,35 @@ function write_tcp(spec, data) {
     view.setUint8(12, doff << 4);
     view.setUint8(13, flags);
     view.setUint16(14, tcp.winsize);
-    view.setUint16(16, 0); // Checksum is 0 during calculation
+    view.setUint16(16, 0); // checksum initially zero before calculation
     view.setUint16(18, tcp.urgent || 0);
 
-    let total_len = (doff * 4) + (spec.tcp_data ? spec.tcp_data.length : 0);
-
-    let checksum = 0;
-    let psudo_header = new Uint8Array(12);
-    let phview = new DataView(psudo_header.buffer, psudo_header.byteOffset, psudo_header.byteLength);
-    for(let i = 0; i < 4; ++i) {
-        phview.setUint8(i, spec.ipv4.src[i]);
-        phview.setUint8(4 + i, spec.ipv4.dest[i]);
-    }
-    phview.setUint8(9, IPV4_PROTO_TCP);
-    phview.setUint16(10, total_len);
-
-    for(let i = 0; i < 6; ++i) {
-        // TODO(perf)
-        checksum += phview.getUint16(i << 1);
-        if(checksum > 0xFFFF) {
-            checksum = (checksum & 0xFFFF) + 1;
-        }
-    }
-    for(let i = 0; i < doff * 2; ++i) {
-        checksum += view.getUint16(i << 1);
-        if(checksum > 0xFFFF) {
-            checksum = (checksum & 0xFFFF) + 1;
-        }
-    }
+    const total_length = TCP_HEADER_SIZE + (spec.tcp_data ? spec.tcp_data.length : 0);
 
     if(spec.tcp_data) {
-        for(let i = 0; i < spec.tcp_data.length; i += 2) {
-            checksum += spec.tcp_data[i] << 8 | spec.tcp_data[i+1];
-            if(checksum > 0xFFFF) {
-                checksum = (checksum & 0xFFFF) + 1;
-            }
-        }
+        view.setArray(TCP_HEADER_SIZE, spec.tcp_data);
     }
-    view.setUint16(16, checksum ^ 0xFFFF);
 
-    return doff * 4;
-}
-
-function make_packet(spec) {
-    // TODO: Can we reuse this buffer?
-    let bytes = new Uint8Array(1518); // Max ethernet packet size
-    dbg_assert(spec.eth);
-
-    let written = write_eth(spec, bytes);
-    return bytes.subarray(0, written);
+    const pseudo_header =
+        (spec.ipv4.src[0] << 8 | spec.ipv4.src[1]) +
+        (spec.ipv4.src[2] << 8 | spec.ipv4.src[3]) +
+        (spec.ipv4.dest[0] << 8 | spec.ipv4.dest[1]) +
+        (spec.ipv4.dest[2] << 8 | spec.ipv4.dest[3]) +
+        IPV4_PROTO_TCP +
+        total_length;
+    view.setInetChecksum(16, total_length, pseudo_header);
+    return total_length;
 }
 
 function fake_tcp_connect(dport, adapter)
 {
-    // TODO: check port collisions
-    let sport = 49152 + Math.floor(Math.random() * 1000);
-    let tuple = [
-        adapter.vm_ip.join("."),
-        dport,
-        adapter.router_ip.join("."),
-        sport
-    ].join(":");
+    const vm_ip_str = adapter.vm_ip.join(".");
+    const router_ip_str = adapter.router_ip.join(".");
+    let sport, tuple;
+    do {
+        sport = 1000 + Math.random() * 64535 | 0;
+        tuple = `${vm_ip_str}:${dport}:${router_ip_str}:${sport}`;
+    } while(adapter.tcp_conn[tuple]);
```

Comment:
I'd suggest something along the lines of (untested):

```javascript
let sport = 1000 + Math.random() * 64535 | 0;
for(let i = 0; i < 65536; i++)
{
    tuple = `${vm_ip_str}:${dport}:${router_ip_str}:${sport}`;
    if(!adapter.tcp_conn[tuple]) break;
    sport = Math.min(1000, (sport + 1) % 65536);
}
if(adapter.tcp_conn[tuple])
{
    dbg_log("warning: no free ports");
    return;
}
````

To prevent infinite loops if ports are exhausted.

Reply:
I rewrote that as you suggested, I had also introduced a wrong dynamic port range and changed it to the proper range I found in RFC6335.

I however think the right thing to do when the pool of dynamic TCP ports is exhausted is to throw an Exception to inform the caller about it (to teardown the connection), if you don't mind (even though complete error handling is still TODO).